src/metadata/metadata_passive_io.c

```diff
@@ -0,0 +1,377 @@
+		/* Moving cline from the freelist to the default partitioin */
+		remove_from_freelist(cache, cline);
+
+		ocf_hb_cline_prot_lock_wr(&cache->metadata.lock, lock_idx, core_id,
+				core_line);
+		OCF_METADATA_LRU_WR_LOCK(cline);
+		ocf_cline_rebuild_metadata(cache, core_id, core_line, cline);
+		OCF_METADATA_LRU_WR_UNLOCK(cline);
+		ocf_hb_cline_prot_unlock_wr(&cache->metadata.lock, lock_idx, core_id,
+				core_line);
+		ocf_cleaning_init_cache_block(cache, cline);
+		ocf_cleaning_set_hot_cache_line(cache, cline);
+
+	} else {
+		/* Cline stays on the freelist*/
+
+		/* To prevent random values in the metadata fill it with the defaults */
+		metadata_init_status_bits(cache, cline);
+		ocf_metadata_set_core_info(cache, cline, OCF_CORE_MAX, ULLONG_MAX);
+	}
+}
+
+static void handle_previously_valid(ocf_cache_t cache,
+		ocf_cache_line_t cline)
+{
+	ocf_core_id_t core_id;
+	uint64_t core_line;
+	uint32_t lock_idx = ocf_metadata_concurrency_next_idx(cache->mngt_queue);
+
+	ocf_metadata_get_core_info(cache, cline, &core_id, &core_line);
+
+	if (metadata_test_dirty(cache, cline) && core_id < OCF_CORE_MAX) {
+		/* Cline stays on the default partition*/
+		ENV_BUG_ON(!metadata_test_valid_any(cache, cline));
+
+		remove_from_default(cache, cline);
+
+		ocf_hb_cline_prot_lock_wr(&cache->metadata.lock, lock_idx, core_id,
+				core_line);
+		OCF_METADATA_LRU_WR_LOCK(cline);
+		ocf_cline_rebuild_metadata(cache, core_id, core_line, cline);
+		OCF_METADATA_LRU_WR_UNLOCK(cline);
+		ocf_hb_cline_prot_unlock_wr(&cache->metadata.lock, lock_idx, core_id,
+				core_line);
+		ocf_cleaning_set_hot_cache_line(cache, cline);
+
+	} else {
+		/* Moving cline from the default partition to the freelist */
+		ocf_cleaning_purge_cache_block(cache, cline);
+		_reset_cline(cache, cline);
+	}
+}
+
+static inline void update_list_segment(ocf_cache_t cache,
+		ocf_cache_line_t begin, ocf_cache_line_t end)
+{
+	ocf_cache_line_t cline;
+
+	/* The last page of collision section may contain fewer entries than
+	   entries_in_page */
+	end = OCF_MIN(end, ocf_metadata_collision_table_entries(cache));
+
+	for (cline = begin; cline < end; cline++) {
+		ocf_part_id_t part_id;
+
+		metadata_clear_dirty_if_invalid(cache, cline);
+		metadata_clear_valid_if_clean(cache, cline);
+
+		part_id = ocf_metadata_get_partition_id(cache, cline);
+		switch (part_id) {
+			case PARTITION_FREELIST:
+				handle_previously_invalid(cache, cline);
+				break;
+			case PARTITION_DEFAULT:
+				handle_previously_valid(cache, cline);
+				break;
+			default:
+				ocf_cache_log(cache, log_crit, "Passive update: invalid "
+						"part id for cacheline %u: %hu\n", cline, part_id);
+				ENV_BUG();
+				break;
+		}
+	}
+}
+
+static void _dec_core_stats(ocf_core_t core)
+{
+	ocf_part_id_t part = PARTITION_DEFAULT;
+
+	ENV_BUG_ON(env_atomic_dec_return(&core->runtime_meta->cached_clines) < 0);
+	ENV_BUG_ON(env_atomic_dec_return(
+				&core->runtime_meta->part_counters[part].cached_clines) < 0);
```

Comment:
lines too long

Reply:
It's probably caused by tab width equal to 8 :) with vim configured to `tabstop=4` it fits 80 characters limit